[RNMobile] Search Block: Add UI Tests and make available to production!

AmandaRiu opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->
**Gutenberg Mobile PR:** https://github.com/wordpress-mobile/gutenberg-mobile/pull/3210


## Description
<!-- Please describe what you have changed or added -->
This PR adds automated UI tests for the Search block and **is the final PR in a series to port the Search block to mobile**. Includes the following:
- Takes the Search Block out of `devOnly` mode. 
- Adds `testID` properties to the Search block label, input and button elements so they can be accessed reliably by Appium when running on iOS. 
- Adds unit tests to ensure `testID`'s exist so they are available later for UI testing. Also updated the snapshots to include these new properties. 
- Adds UI tests for modifying the text of Search block elements. 
- Adds UI tests for the various search block settings options. 

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->
#### Test this PR by first running the unit test:
```
> npm run test-unit:native -- packages/block-library/src/search/test/edit.native.js
```

#### Then run the UI tests for this block:

**IOS**
```
> npm run native test:e2e:bundle:ios
> TEST_RN_PLATFORM=ios npm run native device-tests:local gutenberg-editor-search.test.js
```

**Android**
```
> npm run native test:e2e:bundle:android
> npm run native device-tests:local gutenberg-editor-search.test.js
```

## Screenshots <!-- if applicable -->
N/A

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->
New feature

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

Commit:
Add initial basic UI test for search block
- Had to take the search block out of devOnly mode for this to work.

Commit:
Search block UI tests -- wip

Commit:
Work in progress

Commit:
Add testIDs for iOS UI tests

Commit:
Simplify access to block child components

Commit:
Finished iOS and Android tests for editing child elements

Commit:
Add method to open block settings and dismiss

Commit:
Add Search block settings UI tests

Commit:
Clean up and optimize Search block tests

Commit:
Add unit tests to check for testIDs

Commit:
Move button below input if width greater than 50%

- Re-evaluate during orientation changes as well

Commit:
Use long button styles only when container or button width changes

Commit:
Update unit-test snapshots

Commit:
Remove unused code

Commit:
Merge branch 'rnmobile/search-block-long-button-text' into rnmobile/search-block-ui-tests

Commit:
Update unit-test snapshot

Commit:
Add changelog note for Search Block release

Commit:
Merge branch 'trunk' into rnmobile/search-block-ui-tests

Commit:
Updated unit-test snapshots for changes in long text

dcalhoun commented on packages/block-library/src/search/test/edit.native.js:
```diff
@@ -59,6 +59,35 @@ describe( 'Search Block', () => {
+				instance.findByProps( { testID: 'search-block-button' } );
+			} ).not.toThrow();
+		} );
+	} );
```
>Adds unit tests to ensure `testID`'s exist so they are available later for UI testing. Also updated the snapshots to include these new properties.

I _think_ I understand the intentions here. However, having tests to ensure other tests pass feels a bit excessive. I think each test we add comes with a cognitive and practical "cost" — the price to run and maintain it. I question if the value of these tests outweighs that cost. 

I did see the discussion in #18832, but I feel like that was a very unique circumstance. To my understanding the `testID` added there was in service of tests that reside _outside_ of the `gutenberg` project. So, having unit tests to verify the `testID` remained in place makes sense to me. 

What do you think?

AmandaRiu replied:
That all sounds very reasonable to me and I had wondered the same. Reverted the tests in fe0a2d2c8fe11b47c2cceb8f2bc3a46e4a88ecb1

Commit:
Merge branch 'trunk' into rnmobile/search-block-ui-tests

dcalhoun commented on packages/react-native-editor/CHANGELOG.md:
```diff
@@ -11,6 +11,8 @@ For each user feature we should also add a importance categorization label  to i
 
 ## Unreleased
 
+-   [*] Image block: Improve text entry for long alt text. [#29670]
```
I believe this may have been erroneously added in a merge conflict resolution. It is already present in `1.51.0` in this file. 
```suggestion
```

dcalhoun commented on packages/react-native-editor/__device-tests__/pages/editor-page.js:
```diff
@@ -570,6 +587,57 @@ class EditorPage {
+			locator = `//android.view.View[starts-with(@content-desc, "${ testKeys.android }")]/android.widget.EditText`;
+		}
+
+		return await this.driver.elementByXPath( locator );
```
This may also relate to the discussion in #18832, but is there a specific reason for the tests in this PR to leverage XPaths rather than an accessibility IDs? 

In my opinion, authoring tests from the perspective of the user can be more beneficial rather than finding elements by internal IDs. I do recall your post when you outlined the challenges you faced in locating elements with accessibility IDs. I wonder if a strategy of [finding the _last_ element with the accessibility ID](https://github.com/WordPress/gutenberg/blob/40cbf27d262f474c52dcc27ffc30c25e8256a943/packages/react-native-editor/__device-tests__/gutenberg-editor-cover.test.js#L67) might be applicable?

> XPath: Search the app XML source using xpath (not recommended, has performance issues)

> Try not to use XPath locators unless there is absolutely no other alternative. In general, xpath locators might be many times slower than other types of locators like accessibility id, class name and predicate (up to 100 times slower in some special cases). They are so slow, because xpath location is not natively supported by Apple's XCTest framework.

Additionally, my understanding is that [XPath locators](https://appium.io/docs/en/commands/element/find-elements/index.html#selector-strategies) can be much slower and are [not recommended](https://appium.io/docs/en/advanced-concepts/migrating-to-xcuitest/#xpath-locator-strategy) within the Appium documentation. Please keep in mind, I recognize a lot of the existing e2e tests we have rely upon XPath locators, so I raise this point in hopes of prompting productive discussion, not to "call anyone out." 😄 

WDYT?

AmandaRiu replied:
All very good points. The problem with `elementByAccessibilityId` is that it's an *exact match* locator strategy, but the `accessibilityLabel`s used in the Search block are highly customized for an optimal screenreader experience. This means the label is dynamic and changes depending on the value of the field. Using this method would require the tests to always know the value of each element in the block which could lead to flaky or easily broken tests later down the road. 

To address this I decided to change it up a bit. I added some extra logic to the main search block code to set `accessibilityLabel` to a testID when a screenreader is not actively enabled (ddc4f3dfaf499abbfc1d68cc44a111c1a7c4f4d0 and ddc4f3dfaf499abbfc1d68cc44a111c1a7c4f4d0). I also had to clear out `accessibilityHint` since that gets appended to the `accessibilityLabel` and then used as the `content-desc` in Android. Unfortunately `accessibilityLabel` can only be set on the parent container (`<View/>`) so in order to update the text of an element during Android tests I used `elementByClassName()` which helped to grab the child element needed. Tests updated in e4f0aea48727406195861ea4cf4c782836589f1e. 

The settings portion of the tests I've decided to leave as-is since my code doesn't control the `accessibilityLabel` and it too is dynamic and includes the value of the setting at the time of access. I didn't want to introduce complexity that would later lead to issues. I did some research and it looks like once we upgrade `react-native` to > 0.65, `testID` will work on both Android and iOS which will simplify access to elements without worrying about effecting accessibility. It would probably make more sense to then go back and update the code.

dcalhoun reviewed (commented):
Thank you for taking the time to add such great test coverage to our product. Very excited about this work!

Commit:
Remove unit-tests for testIDs

AmandaRiu committed:
Update packages/react-native-editor/CHANGELOG.md

Co-authored-by: David Calhoun <dpcalhoun@gmail.com>

Commit:
Track when a screenreader is active

Commit:
Set accessibilityLabel to match testID when screenreader is not active

Commit:
Access search block elements by accessibilityID

Commit:
Update unit test snapshot

AmandaRiu commented:
Ready for another review!

AmandaRiu commented:
Special Note: Since the last few commits made changes to the language of the `accessibilityLabel` which would effect how a  screenreader may work, I tested VoiceOver and Talkback on physical devices to ensure no functionality was negatively affected.

dcalhoun commented on packages/block-library/src/search/edit.native.js:
```diff
@@ -204,24 +235,68 @@ export default function SearchEdit( {
+	 * @return {string} The accessibilityLabel for the Search Button
+	 */
+	const getAccessibilityLabelForButton = () => {
+		if ( ! isScreenReaderEnabled ) {
+			return 'search-block-button';
+		}
```
For whatever reason I cannot reply to my original https://github.com/WordPress/gutenberg/pull/30783#discussion_r616986642. So, I'll revive that discussion here a bit as it is directly related to these new lines of code. 

I definitely relate to and recognize cross-platform automated e2e tests are difficult, but setting an accessibility label purely for tests seems like a dangerous practice to me. In theory a user would never encounter these strings, but the added complexity of toggling the labels and hints makes me worry that maybe one day we would expose them (i.e. have them announced to a user) or merely break the proper labels that should be present. 

>The problem with `elementByAccessibilityId` is that it's an exact match locator strategy, but the `accessibilityLabel`s used in the Search block are highly customized for an optimal screenreader experience. This means the label is dynamic and changes depending on the value of the field. Using this method would require the tests to always know the value of each element in the block which could lead to flaky or easily broken tests later down the road.

Is the flakiness or brittleness of the e2e tests leveraging dynamic accessibility labels something you encountered or just a fear that it may happen? 

>The settings portion of the tests I've decided to leave as-is since my code doesn't control the `accessibilityLabel` and it too is dynamic and includes the value of the setting at the time of access. I didn't want to introduce complexity that would later lead to issues.

That makes sense to me. Refactoring the approach for interfacing with block settings here is likely an endeavor that could uncover a lot more work. 

>I did some research and it looks like once we upgrade `react-native` to > 0.65, `testID` will work on both Android and iOS which will simplify access to elements without worrying about effecting accessibility. It would probably make more sense to then go back and update the code.

Do I understand correctly that whenever we upgrade to > 0.65 the hope would be to replace the test-oriented accessibility labels used for automated tests with `testID`s? Would you be willing to share links you found during your research regarding `testID`s? 

I raise most of the above for my own education. You obviously dove deep into our e2e suite. So, I do not intend to demand refactoring the approach, but to learn about the caveats you discovered and the bests approaches to address them. I look forward to your thoughts on these topics. Thank you for taking me along on the journey! 🙇🏻

dcalhoun replied:
I appreciate the insight you have posted on internal channels, much of which addressed a lot of my previous comment. Please feel free to not rehash anything you believe you have already addressed elsewhere.

AmandaRiu replied:
Thanks for the review @dcalhoun ! I'm going to answer this a little out of order so bear with me. 

> Is the flakiness or brittleness of the e2e tests leveraging dynamic accessibility labels something you encountered or just a fear that it may happen?

Fear that it may happen because it happened to me the first time around when I tried moving over to using `accessibilityID`'s and I had to back out of those changes. 

> Do I understand correctly that whenever we upgrade to > 0.65 the hope would be to replace the test-oriented accessibility labels used for automated tests with testIDs? Would you be willing to share links you found during your research regarding testIDs?

Yup, you understand it correctly. [I read about it in this blog post](https://dev.to/nextlevelbeard/an-end-to-the-abuse-of-accessibility-ids-5d2j) and the [actual PR that implemented this fix can be found here](https://github.com/facebook/react-native/pull/29610). 

> I definitely relate to and recognize cross-platform automated e2e tests are difficult, but setting an accessibility label purely for tests seems like a dangerous practice to me.

It's definitely not ideal just like writing Appium tests around the `accessibilityLabel` is not ideal. That said, the only way the users would encounter these alternate strings is if they were inspecting the app using something like appium because the moment a screenreader is enabled those alternate "testID" strings are no longer used. That said, I completely understand your concern and if you feel strongly about it I can back out of these changes and go back to using the `elementByXPath` locator strategy and once we upgrade React Native we can go back and optimize the e2e tests using only the `testID`. I'm fine with either approach. WDYT?

AmandaRiu replied:
Additional note about `testID` and RN - the latest version of RN is 0.64 (0.65 hasn't been cut yet) -- so it's unknown when that fix to allow Appium to access `testID` on Android will be released. [The commit to add that feature](https://github.com/facebook/react-native/commit/381fb395ad9d2d48717a5d082aaedbecdd804554) is currently in master so it _should_ make the next release cut.

dcalhoun replied:
>I'm fine with either approach. WDYT?

Thank you for taking the time to consider my thoughts and questions. I believe retaining the current approach is likely the best path forward.

dcalhoun commented on packages/block-library/src/search/test/__snapshots__/edit.native.js.snap:
```diff
@@ -7,10 +7,11 @@ exports[`Search Block renders block with button inside option 1`] = `
   <View
-    accessibilityHint="Double tap to edit label text"
-    accessibilityLabel="Search block label. Current text is Search"
+    accessibilityHint={null}
```
If desired, we could set the `null` values to `undefined` to avoid having them show up in the snapshots entirely. I think either value is fine and don't have a strong preference.

AmandaRiu replied:
Sweet! TIL'd that was possible. Updated in 2e7a4efa51ea8de248fd94e0badf40a5a154eb4d

dcalhoun commented on packages/react-native-editor/__device-tests__/pages/editor-page.js:
```diff
@@ -570,6 +583,64 @@ class EditorPage {
+		const textViewElement = await this.getSearchBlockTextElement( testID );
+
+		if ( isAndroid() ) {
+			// await doubleTap( this.driver, textViewElement );
```
Was the desire to remove this line or do we need to reinstate it?

AmandaRiu replied:
Ugh, good catch. The desire was to remove it...an artifact leftover from my different attempts at getting to the `EditText` element on Android. Removed in e2e5a204471245a4db3a60a02b0ed11f1bcf03ee

dcalhoun commented on packages/block-library/src/search/edit.native.js:
```diff
@@ -204,24 +235,68 @@ export default function SearchEdit( {
+						: null
+				}
+				accessibilityLabel={ getAccessibilityLabelForPlaceholder() }
+				testID="search-block-input"
```
Is it accurate to say the `testID`s in this file are unused until `react-native` > 0.65 plus refactoring or am I misunderstanding?

AmandaRiu replied:
That's a good point. I've removed `testID` now that it's not being used in 17a78d9fcfd60f5716dc00034e1ae42a373dc394. We can add it back when we upgrade RN.

dcalhoun reviewed (commented):
>Special Note: Since the last few commits made changes to the language of the `accessibilityLabel` which would effect how a screenreader may work, I tested VoiceOver and Talkback on physical devices to ensure no functionality was negatively affected.

I also tested this round of changes on an iPhone SE and Samsung Galaxy S20. Both performed as expected, aside from one issue that is related to an issue we discussed once before — the auto-focus of the search input. I wanted to at least showcase the issue I encountered incase we want to capture it in separate issue. The fact that the erroneous search input auto-focus breaks the focus of the bottom sheet makes this a bigger issue from my perspective. 

_Steps to Reproduce_

1. Launch block editor. 
2. Add Search block. 
3. Enable VoiceOver. 
4. Open block settings. 

_Expected Outcome:_ The bottom sheet is focused, and focus is trapped within the bottom sheet modal until it is dismissed. 

_Actual Outcome:_ The search input is focused. A user would have to known the bottom sheet is open and tap it to move focus to its elements, which may be difficult to do if one is visual impaired. 

<details>
<summary>Screenshot of issue</summary>

https://user-images.githubusercontent.com/438664/115619032-087d2200-a2b9-11eb-8346-86ec94d98d07.MP4

</details>

Commit:
Use undefined instead of null so it doesn't show up in snapshot

Commit:
Remove unused code

Commit:
Remove testID as no longer needed for iOS e2e

AmandaRiu commented:
@dcalhoun 

Thanks for the additional review. I've answered all the code comments inline and made some changes in response. Regarding: 

> I also tested this round of changes on an iPhone SE and Samsung Galaxy S20. Both performed as expected, aside from one issue that is related to an issue we discussed once before — the auto-focus of the search input. I wanted to at least showcase the issue I encountered incase we want to capture it in separate issue. The fact that the erroneous search input auto-focus breaks the focus of the bottom sheet makes this a bigger issue from my perspective. [...]

I tested this as well, but I do not believe this is an issue with the Search block directly. Opening any block's settings while a screenreader is active does not place the focus in the bottom sheet. The user is forced to swipe to navigate to it. I think this is likely a larger problem with maybe the way the bottom sheet is displayed and beyond the scope of this PR. 

Ready for another round, thank you!

dcalhoun reviewed (approved):
This looks good! :shipit: 

>I tested this as well, but I do not believe this is an issue with the Search block directly. Opening any block's settings while a screenreader is active does not place the focus in the bottom sheet. The user is forced to swipe to navigate to it. I think this is likely a larger problem with maybe the way the bottom sheet is displayed and beyond the scope of this PR.

Ah, thank you for verifying this. Definitely seems unrelated to this work.

Commit:
Remove forced clear text option and expand sleep
Since the search block fields are already empty, accept whatever the
default value for `clear` which is `true` for Android and `false` for
iOS. Also add a sleep when changing animated settings options on Android to
avoid flakiness.

Commit:
Avoid test flakiness by making search block fields
empty before adding text.

AmandaRiu commented:
Made some additional improvements to the UI tests after finding they were failing on iOS only when run from CircleCI. For some reason the test wasn't getting properly replaced so the expect() statements were failing. To get around this odd flakiness that doesn't happen when run locally for anyone who has tested this PR, I've changed the set of tests around modifying the text in elements of the Search block. For those tests I now manually insert a Search block with all text elements blank. This way it doesn't matter if the existing text is replaced properly or not because there is no existing text. I also hardened the tests so if one of the text-tests fail it doesn't fail them all. 

I've verified these tests now run successfully not only locally, but also on CircleCI. Ready for a final round @dcalhoun !

dcalhoun reviewed (approved):
LGTM. Re-ran the latest e2e tests. Tested on an iPhone SE and Samsung Galaxy S20.
